Add new module "argparse" that extends the class ArgumentParser

JAlvarezJarreta opened the pull request:
The idea is to include useful functionality besides wrapper methods and default behaviour for common scenarios in Ensembl repositories.

MatBarba commented on src/python/ensembl/utils/argparse.py:
```diff
@@ -0,0 +1,142 @@
+    return src_path
+
+
+def _validate_dst_path(parser: argparse.ArgumentParser, dst_path: os.PathLike) -> Path:
+    """Returns the path if it is writable, raises an error through the argument parser otherwise.
+    
+    Args:
+        parser: Argument parser.
+        dst_path: File or directory path to check.
+
+    """
+    dst_path = Path(dst_path)
+    if dst_path.exists() and os.access(dst_path, os.W_OK):
+        return dst_path
+    for parent_path in dst_path.parents:
+        if parent_path.exists() and not os.access(parent_path, os.W_OK):
+            parser.error(f"'{dst_path}' is not writable")
+    return dst_path
```
Do we want to add an option to fail if the dst file already exists? Or is that too much to do at this stage

MatBarba reviewed (approved):
This looks useful, thanks

JAlvarezJarreta commented on src/python/ensembl/utils/argparse.py:
```diff
@@ -0,0 +1,129 @@
+                "--src_host", etc.
+
+        """
+        self.add_argument(f"--{prefix}host", required=True, help="Name of the host")
+        self.add_argument(f"--{prefix}port", required=True, type=int, help="Port number")
+        self.add_argument(f"--{prefix}user", required=True, help="User name")
+        self.add_argument(f"--{prefix}password", metavar="PWD", default="", help="Host password")
```
It is a tricky one, particularly with those capital letters involved. Will try to see if there is something smart that can be done. If not, it may stay like this for now.

JAlvarezJarreta commented on src/python/ensembl/utils/argparse.py:
```diff
@@ -0,0 +1,142 @@
+
+        """
+        self.add_server_arguments(prefix=prefix)
+        self.add_argument(f"--{prefix}database", required=True, metavar="NAME", help="Database name")
```
Opted for a grouping option to ease help readability.

JAlvarezJarreta commented:
Addressed all comments. FYI, I have opted for inheritance rather than encapsulation because we still want to provide the default functionality for any current and future methods `argparse.ArgumentParser` may provide, so people can still use it in the same way.

marcoooo commented on src/python/ensembl/utils/__init__.py:
```diff
@@ -12,5 +12,3 @@
 # limitations under the License.
 """Utils module"""
-
-from .rloader import *
```
you just broke the handover deployment while doing that.

marcoooo replied:
No... in fact your didn't but you could have :-) 
- https://github.com/Ensembl/ensembl-prodinf-datachecks/blob/bbca777d04b9d2dac99acdb8b6a3c3efef75d641/src/ensembl/production/datacheck/config.py#L22
- https://github.com/Ensembl/ensembl-prodinf-handover/blob/1d384cd212d3107e0c1af850caf025486c1c70db/src/ensembl/production/handover/config.py#L22

JAlvarezJarreta replied:
Phew! Will be more cautious in the future for sure! :eyes:

marcoooo commented:
Please just resolve the last bit after reading.
